Bad instruction in test xexpression.shared_xfunctions on MacOS 10.13 XCode 9.4.1

When I run the tests from latest xtensor on MacOS 10.13 with XCode 9.4.1, I get the below error. I'm not really sure how to isolate if my -march-settings are too aggressive for this old Xeon processor, or if there is an actual bug. Help would be appreciated!

```
[ RUN      ] xexpression.shared_xfunctions
Process 22298 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x0000000100148a24 test_xtensor_lib`thread-local wrapper routine for xt::detail::flat_expression_adaptor<xt::xfunction<xt::detail::plus, xt::xshared_expression<xt::xarray_container<xt::uvector<double, xsimd::aligned_allocator<double, 16ul> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >, xt::xshared_expression<xt::xarray_container<xt::uvector<double, xsimd::aligned_allocator<double, 16ul> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> > > const, (xt::layout_type)1>::m_index + 4
test_xtensor_lib`thread-local wrapper routine for xt::detail::flat_expression_adaptor<xt::xfunction<xt::detail::plus, xt::xshared_expression<xt::xarray_container<xt::uvector<double, xsimd::aligned_allocator<double, 16ul> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >, xt::xshared_expression<xt::xarray_container<xt::uvector<double, xsimd::aligned_allocator<double, 16ul> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> > > const, (xt::layout_type)1>::m_index:
->  0x100148a24 <+4>: ud2    
    0x100148a26 <+6>: nopw   %cs:(%rax,%rax)

test_xtensor_lib`xt::detail::flat_expression_adaptor<xt::xfunction<xt::detail::plus, xt::xshared_expression<xt::xarray_container<xt::uvector<double, xsimd::aligned_allocator<double, 16ul> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >, xt::xshared_expression<xt::xarray_container<xt::uvector<double, xsimd::aligned_allocator<double, 16ul> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> > > const, (xt::layout_type)1>::~flat_expression_adaptor:
    0x100148a30 <+0>: pushq  %rbp
    0x100148a31 <+1>: movq   %rsp, %rbp
```

It may a bit more changes, because with only this line changed I get:
```
xtensor/include/xtensor/xstrided_view_base.hpp:77:106: error: 'xt::detail::flat_expression_adaptor<CT, L>::index_type xt::detail::flat_expression_adaptor<CT, L>::m_index' is not a static data member of 'class xt::detail::flat_expression_adaptor<CT, L>'
         thread_local typename flat_expression_adaptor<CT, L>::index_type flat_expression_adaptor<CT, L>::m_index;
                                                                                                          ^~~~~~~
xtensor/include/xtensor/xstrided_view_base.hpp:77:106: error: template definition of non-template 'xt::detail::flat_expression_adaptor<CT, L>::index_type xt::detail::flat_expression_adaptor<CT, L>::m_index'
```

--- a/include/xtensor/xstrided_view_base.hpp
+++ b/include/xtensor/xstrided_view_base.hpp
@@ -66,14 +66,12 @@ namespace xt
 
         private:
 
+            static index_type& get_index();
+
             mutable CT* m_e;
             inner_strides_type m_strides;
-            thread_local static index_type m_index;
             size_type m_size;
         };
-
-        template <class CT, layout_type L>
-        thread_local typename flat_expression_adaptor<CT, L>::index_type flat_expression_adaptor<CT, L>::m_index;
 
         template <class T>
         struct is_flat_expression_adaptor : std::false_type
@@ -678,7 +676,7 @@ namespace xt
         inline flat_expression_adaptor<CT, L>::flat_expression_adaptor(CT* e)
             : m_e(e)
         {
-            resize_container(m_index, m_e->dimension());
+            resize_container(get_index(), m_e->dimension());
             resize_container(m_strides, m_e->dimension());
             m_size = compute_size(m_e->shape());
             compute_strides(m_e->shape(), L, m_strides);
@@ -689,7 +687,7 @@ namespace xt
         inline flat_expression_adaptor<CT, L>::flat_expression_adaptor(CT* e, FST&& strides)
             : m_e(e), m_strides(xtl::forward_sequence<inner_strides_type, FST>(strides))
         {
-            resize_container(m_index, m_e->dimension());
+            resize_container(get_index(), m_e->dimension());
             m_size = m_e->size();
         }
 
@@ -710,16 +708,16 @@ namespace xt
         inline auto flat_expression_adaptor<CT, L>::operator[](size_type idx) -> reference
         {
             auto i = static_cast<typename index_type::value_type>(idx);
-            m_index = detail::unravel_noexcept(i, m_strides, L);
-            return m_e->element(m_index.cbegin(), m_index.cend());
+            get_index() = detail::unravel_noexcept(i, m_strides, L);
+            return m_e->element(get_index().cbegin(), get_index().cend());
         }
 
         template <class CT, layout_type L>
         inline auto flat_expression_adaptor<CT, L>::operator[](size_type idx) const -> const_reference
         {
             auto i = static_cast<typename index_type::value_type>(idx);
-            m_index = detail::unravel_noexcept(i, m_strides, L);
-            return m_e->element(m_index.cbegin(), m_index.cend());
+            get_index() = detail::unravel_noexcept(i, m_strides, L);
+            return m_e->element(get_index().cbegin(), get_index().cend());
         }
 
         template <class CT, layout_type L>
@@ -753,9 +751,16 @@ namespace xt
         }
 
         template <class CT, layout_type L>
-        inline auto flat_expression_adaptor<CT, L>::cend() const ->const_iterator
+        inline auto flat_expression_adaptor<CT, L>::cend() const -> const_iterator
         {
             return m_e->template cend<L>();
+        }
+
+        template <class CT, layout_type L>
+        inline auto flat_expression_adaptor<CT, L>::get_index() -> index_type&
+        {
+            thread_local static index_type index;
+            return index;
         }
     }
 

